### PR TITLE
Release first working memtester diag

### PR DESCRIPTION
**Pull request by Goshik92**
This PR contains the following features:
- OCP-compliant memtester diag
- Lex/Yacc-based memtester output parser
- Unit tests for the parser
- Docker image for testing the diag with different memtester versions
- Basic documentation

The following features are not a part of this PR and may or may not be introduced in later PRs:
- CD/CI
- reporting memtester version in DUT info
- output validation for OCP compliance
- unit tests for the entire diag
- diag testing based on error injection
- guide on developing diags using PLY

### CODE REVIEW

File: README.md
Comment (mimir-d) [inline README.md:9]: i'll add some more "release" type instructions here later on (different PR). these current instructions are more geared towards devs

File: README.md
Comment (mimir-d) [inline README.md:37]: we can probably cover these with gh actions to make sure they continue to work on all PRs (memtester bin should work in the gh containers or at least produce some relevant output even if it doesnt have access to raw memory)

File: Dockerfile
Comment (mimir-d) [inline Dockerfile:12]: this implies x64 right? we should make this compile from source longer term since running upstream prebuilt may be a security issue for some people

File: README.md
Comment (mimir-d) [inline README.md]: would remove this; plus the general recommendation is to add the user to the `docker` group, rather than sudo the commands

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py:57]: what is the minimum python version for this? `list[Union[...]]` is somewhat new for some users (as opposed to `typing.List`)
min version should also be in the readme

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py:14]: would be nice to have some docstring for these dataclasses; maybe with sample input

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:57]: there's some current users im aware of that would need py3.7 and `list[...]` is not supported in that version iirc

File: README.md
Comment (mimir-d) [inline README.md:54]: some instructions for developers on typechecking, linter, etc? doesnt necessarily have to be in this PR, so you could track it as an issue otherwise

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py:115]: can these have type annotations too?

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py]: nitpicking; `FLR` doesnt really read as failure; maybe just use `FAIL`?

File: src/main.py
Comment (mimir-d) [inline src/main.py:30]: arguably this should use the memtester version here

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py]: nitpicking; usually not a good idea to add individual people names here because this list can grow indefinitely

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py]: this should be `OCP Test & Validation` (or add the `Test & Validation` part to existing line)

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py:79]: what is the rationale of using a dataclass with callbacks here instead of a normal class with overridable functions?

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py]: nitpicking; stray `;` (that i also have a habit of from c/cpp)

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py]: is there a risk of `want|got` to incorrectly ignore some lines?

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py]: what type is `stdout_gen` here?

File: src/memtester_parsing.py
Comment (mimir-d) [inline src/memtester_parsing.py:219]: all fields of `self.callbacks` are None here; so if this `MemtesterObserver` were to be used as it's constructed here, any caller would fail while trying to call `None` callables. As such, this object instantiation here is incomplete

we should have "correct by construction` semantics for all of the objects in this codebase. this correlates with my prev comment on the callbacks dataclass. if that was a normal class with default impls for the callback methods, this instance here would be correct

File: src/main.py
Comment (mimir-d) [inline src/main.py:55]: id argue the cardinality of `loop` here should be equal to `tv.step`, but it could also be that a step is iterating thru a memory zone multiple times. opinion?

File: src/main.py
Comment (mimir-d) [inline src/main.py:43]: i think this behavior should be gated by a program arg; to let the user decide if they accept the program possibly failing with errors or refuse to run a nonsupported memtester bin from the start

File: src/main.py
Comment (mimir-d) [inline src/main.py:40]: the determination of whether this version is acceptable should be done before the `tv.run` begins. This does indeed result in running the binary twice (first with `--version`) but i find that acceptable and this
1. allows the user to say "i refuse to run this if not a supported version"
2. the version of memtester to be correctly reported in dut.SoftwareInfos, rather than a log artifact

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py]: maybe you want a `^` at the start of the regex here

File: src/main.py
Comment (mimir-d) [inline src/main.py]: note that measurement series has the semantic of chronological data (spec: `A measurement series consists of time based data`), which doesnt appear to be your case here

Comment (mimir-d) [reply src/main.py]: also should add the unit of measurement (but this will apply to the `loop_callback` code)

File: src/main.py
Comment (mimir-d) [inline src/main.py]: nitpick; would say here `Memory test {} passed`, to be more explicit what kind of test

File: src/main.py
Comment (mimir-d) [inline src/main.py]: it's not clear from this that the diag wrapper here **will write** to this file (i first read this as if it were input instead of running the memtester binary)

File: src/main.py
Comment (mimir-d) [inline src/main.py]: does memtester have any args that contain spaces (either in key or value)? if so, this should split incorrectly, eg. `--mt_args "--some-param \"v1 v2\""`

File: src/main.py
Comment (mimir-d) [inline src/main.py:109]: this `sh` module is pretty nice; i'll note that the behavior of `_iter` couldve been implemented in a pretty short manner here as well, avoiding an extra dependency

File: src/main.py
Comment (mimir-d) [inline src/main.py:112]: i mentioned before, but this place is a bit far from the other comment; i think `memtester-not-found` is not a `tv.step` problem, but rather something that should be detected before starting any `tv.run`s

**Review by mimir-d — CHANGES_REQUESTED** [review]
overall pretty good; some things mandatory to address

File: src/main.py
Comment (mimir-d) [reply src/main.py:30]: im gonna close this comment; you say this is not in scope in the description above
> reporting memtester version in DUT info

File: Dockerfile
Comment (Goshik92) [reply Dockerfile:12]: This is already being compiled from source. Notice the `make` command at line 12. Or maybe I misunderstood the question...

File: README.md
Comment (Goshik92) [reply README.md]: Apparently, "the docker group grants root-level privileges to the user". Added a link to the official guide instead of giving the reader my sudo advice.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py]: Indeed:
image

Done.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py]: Done. We should be careful with moving code when reorganizing repos. Dylan's authorship was erased from some of his code, IIRC.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py]: Done.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:79]: I guess this is just a matter of preference. Some people prefer OOP-style (overriding), others like functional style (callbacks). For this particular case, both approaches seem equivalent to me.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py]: Thanks for noticing! Done.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py]: You are right. I had `^` initially, but then removed it for some reason. It is better to have it. Done.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:219]: `MemtesterCallbacks` initializes all its members with [empty lambdas](https://github.com/opencomputeproject/ocp-diag-memtester/blob/7d11a44daad738e603a8b27b201b5be79843482a/src/memtester_parsing.py#L45).

File: src/main.py
Comment (Goshik92) [reply src/main.py:55]: That was my initial thought too. But then I realized that the user may request 1000 iterations, which will generate a thousand test steps. I think something unique rather than generated would better qualify for a test step. Memtester installation and version checks could be separate test steps, IMO.

File: src/main.py
Comment (Goshik92) [reply src/main.py]: Isn't it clear from the context that those are memory tests?

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:57]: > Importing those from typing is deprecated. Due to [PEP 563](https://peps.python.org/pep-0563/) and the intention to minimize the runtime impact of typing, this deprecation will not generate DeprecationWarnings. Instead, type checkers may warn about such deprecated usage when the target version of the checked program is signalled to be Python 3.9 or newer. It’s recommended to allow for those warnings to be silenced on a project-wide basis.

> The deprecated functionality may eventually be removed from the typing module. Removal will occur no sooner than Python 3.9’s end of life, scheduled for October 2025.

https://peps.python.org/pep-0585/

Not sure what to do about this one. Oct 2025 is in two years.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:14]: Done. Not sure about sample input, though. Cannot think of anything illustrative enough to show what these classes are for. Perhaps, the existing unit tests can give the reader some understanding.

File: README.md
Comment (Goshik92) [reply README.md:54]: https://github.com/opencomputeproject/ocp-diag-memtester/issues/2

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:115]: The `t` object in grammar rules is [dynamic](https://github.com/dabeaz/sly/blob/33d4f5afc04c9cae78bd7dc094e08badc738409e/src/sly/yacc.py#L104), so type annotations will not increase code readability. The object's properties and use scenarios are covered in the official SLY doc, but I'll probably try to touch on this topic if I get to write an [intro article](https://github.com/opencomputeproject/ocp-diag-memtester/issues/3).

File: src/main.py
Comment (Goshik92) [reply src/main.py:43]: Since the parser is strict about the text's grammar, shouldn't we assume that if a newer memtester version does not cause parsing errors, most likely it will work fine with this wrapper? In case when the parser fails, the user will definitely know that because of the ERROR artifact. The warning log message will give the user some hint about why they got the parsing error. Opinion?

File: src/main.py
Comment (Goshik92) [reply src/main.py]: >does memtester have any args that contain spaces?

If I understand the usage hint correctly, it does not:

>Usage: memtester [-p physaddrbase [-d device]] <mem>[B|K|M|G] [loops]

But it is better to fix it anyway. Done.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py]: Done.

File: src/main.py
Comment (Goshik92) [reply src/main.py]: Attempted to improve it.

File: README.md
Comment (Goshik92) [reply README.md:9]: Thanks in advance!

File: README.md
Comment (Goshik92) [reply README.md:37]: Right. But we probably need a library for parsing the OCP TV output format to do that. Or maybe we can just parse JSON manually.

File: src/main.py
Comment (Goshik92) [reply src/main.py:40]: https://github.com/opencomputeproject/ocp-diag-memtester/issues/4

File: src/main.py
Comment (Goshik92) [reply src/main.py]: 1. Done.
2. Isn't the address value unitless?

File: src/main.py
Comment (Goshik92) [reply src/main.py:109]: Yes, a separate function could replace the `sh` module. On the other hand, it is nice to have for free [these](https://github.com/opencomputeproject/ocp-diag-memtester/blob/df725fd5aeca842b6b243d0f2aecab8a8b900c9a/src/main.py#L28) one-liners for extracting dut info. I understand that an extra dependency implies that the users need to audit one more library to use this diag, but I would argue that something like `sh` might be universally useful for all wrapper-diags.

File: src/main.py
Comment (Goshik92) [reply src/main.py:112]: I am confused. How do I report an error without starting a `tv.run`? According to the spec:
>Errors can be child artifacts of either TestRunArtifact or TestStepArtifact messages.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:115]: I just realized I misread this code citation. I thought the call was related to the parser, not lexer. So, please ignore my previous comment. 

`t` is always of the `Token` type. Will we benefit from type-annotating this fact?

File: src/main.py
Comment (Goshik92) [reply src/main.py:43]: UPD: As discussed in the meeting, by default, the execution should fail if the version is not marked as supported. This behavior should be overridible with a flag.

Comment (Goshik92) [reply src/main.py:43]: Done.

File: README.md
Comment (mimir-d) [reply README.md:37]: i dont think the check has to cover 100% from the get go, so we could manually parse the json in a simple py script. What i need to check though is if we have enough hardware access in the GH containers to run this and get accurate data. Would be useful at this point to have that OCPTV hardware lab for remote CI.
In any case, action on my part

File: Dockerfile
Comment (mimir-d) [reply Dockerfile:12]: sorry, i wasnt very clear. What i was thinking here is whether this source is trusted enough to download it and compile as is.
The other question is if this only compiles x86_64 in this particular setup or can we also make/have some cross-chain for stuff like aarch64 (context: there's a team interested in running this against mobile devices)

File: README.md
Comment (mimir-d) [reply README.md]: yeah, this is even better; thanks

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:57]: the bigger point is the min python version. Im not particularly interested on how the typing hints syntax looks like, but rather keeping compat with a lower version (as i mentioned, i know of some users interested in this, having a 3.7 requirement). 

More importantly, the supported python versions should be in the readme (and it's gonna be in a pyproject.toml file as well later). So for now, and this PR, id just say to document that and we can fix compatibility later.

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:14]: thanks for fixing, but note on "docstring". what you added are comments (and generally wont show up in `help(x)` calls, though maybe newer python versions do that?). eg.
```python
# comment
class A:
    "docstring"
    
    addr: int
```

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:115]: generally speaking, it would be ideal to have type hints everywhere possible. This gives the best chance of catching bugs (and autocomplete for lang servers), but also general code architecture (because when the types no longer make sense or are conflictual, it's a clear sign that the code design has a problem -- this is a hint in python, other langs cover this by being compiled)

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py]: > Dylan's authorship was erased from some of his code, IIRC.

i wouldnt necessarily say so, that kind of origin is present in the git history. And over time, can you say a particular file was written by one person or another?
Arguably, in case of distribution, where the git history isnt immediately available, we should always have a pointer to origin, and generally speaking the group having worked on said artifact.

I however do not consider myself an authority on this matter. Different people can interepret this differently. Above is my opinion

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:79]: people yes; but in this case python is an OOP-style language and it would be more pythonic to follow that pattern. As a sidenote, this also introduces a lot of cognitive load in the caller context, where you build a `MemtesterCallbacks` object, and then you piecemeal make it useful

LE. removed text about invalid by construction; i missed the empty lambdas in your initial code

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:219]: sorry, i missed the empty lambdas in the initial review; saw `None` and shortcircuited to `x = None` instead of `x = lambda: None`

my other comment about OOP still stands

File: src/main.py
Comment (mimir-d) [reply src/main.py:55]: "memtester installation"? or am i reading that wrong and it's actually "memtester installation check"?

i dont think there's something wrong in having a dynamic number of steps; it doesnt have to be encoded in source code (in that the cardinality of steps is static).

maybe we dont need to do this in this PR. i'd also ask @dframe-google what he thinks

File: src/main.py
Comment (mimir-d) [reply src/main.py:43]: thanks

File: src/main.py
Comment (mimir-d) [reply src/main.py]: > Isn't the address value unitless?

youve a good point there. Should it be "bytes"? im not very convinced by this unit. Your call

File: src/main.py
Comment (mimir-d) [reply src/main.py]: yeah, ive thought of that. But consider a user who inspects these results in some test executor UI, only looking at the measurements, logs, etc without the name of the test (scrolling down the page for example). I think it would be useful in that case to see "memory".
I'm not gonna insist on this.

File: src/main.py
Comment (mimir-d) [reply src/main.py]: > Path to the log file for dumping raw memtester output

yeah, this works for me; thanks

File: src/main.py
Comment (mimir-d) [reply src/main.py:109]: im not disagreeing; we'll see how this goes

File: src/main.py
Comment (mimir-d) [reply src/main.py:112]: here's an example of error before run start (this case shows issue in DUT discovery): https://github.com/opencomputeproject/ocp-diag-core-python/blob/dev/examples/simple.py#L97

File: Dockerfile
Comment (Goshik92) [reply Dockerfile:12]: >whether this source is trusted enough to download it and compile as is

The source is the memtester author's web site. Can it be absolutely trusted? Well, even if the author does not have ill intent, we do not know what security policies they follow to maintain their server. On the other hand, the code is built and run in a Docker container, which provides an extra security layer. Whether Docker itself can be considered secure is a separate topic, but this is what the official guide says:

>Docker containers are, by default, quite secure; especially if you run your processes as non-privileged users inside the container.

>You can add an extra layer of safety by enabling AppArmor, SELinux, GRSEC, or another appropriate hardening system.

As for cross compilation, I fail to see how this can be helpful. The main purpose of this container is to test whether a particular memtester version causes parsing errors. Are we going to see any differences of memtester's output format across different platforms? Well, there is a chance, but I do not remember seeing any platform-dependent code, related to the output in memtester's src. However, if we absolutely need to verify platform compatibility, we can always run this container on the target platform. The [Docker Ubuntu image](https://hub.docker.com/_/ubuntu) seems to support aarch64, so we can find an ARM machine with Docker and try it.

WDYT?

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py]: Git history is good enough. A was referring to a case when a project was moved from one repo to another without keeping  the history. That confused me, because I did not know who the authors were and whom to ask questions about the code. That is why I added myself to the copyright section initially: if the history got erased, people would know whom to talk to. But as long as we maintain the history properly, we do not need to mention authors directly in the code.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:79]: I am fine with the OOP style if it is considered preferable. I was about to go and switch to it, but then I realized that the functional style has one important advantage. Functions can have separate lexical environments, whereas methods of a class have to share the same lexical environment. In practical terms this means that with classic overriding, it will be impossible to move one of the "callbacks" to a different scope. Currently, this functionality is not required, but it may turn useful in the future (e. g. if we decide to introduce new steps that share the same memtester executable run). In that sense, I consider the callback approach more general. Any thoughts?

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:115]: >because when the types no longer make sense or are conflictual, it's a clear sign that the code design has a problem

I prefer statically typed languages over dynamically typed ones for that very reason. However, in this particular case, it's the SLY library who calls those methods, so I fail to see what design problems can be caught by type hinting. I do, however, agree with the autocompletion argument. Done.

UPD: No, not done. `Token` is not importable ([1](https://github.com/dabeaz/sly/blob/33d4f5afc04c9cae78bd7dc094e08badc738409e/src/sly/lex.py#L34), [2](https://github.com/dabeaz/sly/blob/33d4f5afc04c9cae78bd7dc094e08badc738409e/src/sly/__init__.py#L2)).

File: src/main.py
Comment (Goshik92) [reply src/main.py]: I'd keep it unitless because theoretically we can encounter an architecture with word addressing. This is very unlikely, though.

File: src/main.py
Comment (Goshik92) [reply src/main.py]: Done.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:57]: Declared the current state of things for this PR.

File: src/memtester_parsing.py
Comment (Goshik92) [reply src/memtester_parsing.py:14]: Done.

File: src/main.py
Comment (Goshik92) [reply src/main.py:55]: >"memtester installation check"

This one. "check**s**" was supposed to imply this.

File: src/main.py
Comment (Goshik92) [reply src/main.py:112]: Thanks for the link! 

https://github.com/opencomputeproject/ocp-diag-memtester/issues/5

**Review by dframe-google — APPROVED** [review]

File: Dockerfile
Comment (mimir-d) [reply Dockerfile:12]: >we do not know what security policies they follow to maintain their server

yeah, we should probably move to a github source in the future (if available, i havent checked).

> The main purpose of this container is to test whether a particular memtester version causes parsing errors

This wasn't clear to me, but you do have a section on this in the readme, so it's on me. It would be useful though to move this docker file to a separate folder which hints at **testing**. As is, this pattern (with having the dockerfile in root) usually goes to "build this for a release", if that makes sense. (will merge the PR if nothing else needs changing, and this move can be done later; or without a PR, it doesnt require review)

> Are we going to see any differences of memtester's output format across different platforms

not sure but it could theoretically be possible (different memory spaces maybe). But youre saying there's no platform specific code so im gonna scratch this off the list.

File: src/main.py
Comment (mimir-d) [reply src/main.py:55]: yeah, i was reading it wrong (X and Y checks = X check and Y check), didnt see the 's'. I'm gonna close this for now, because the PR has been up for a while, but maybe you can ping @dframe-google on the opinion here.

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:14]: thanks

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:57]: 3.11; thanks

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:79]: i'm not gonna insist on this anymore, just saying that from a pythonic standpoint, this is not generally something you see in codebases. The callback style is very reminiscent of **C**, while python has been from the start an OOP lang.
The scope separation you mention isnt really a huge benefit in my view, nor do I see it as necessary in any future development (or if so, can be done with explicitly shared state inside the **handler** class)

File: src/memtester_parsing.py
Comment (mimir-d) [reply src/memtester_parsing.py:115]: heh.. sad when the lib you wanna use doesnt have typings

**Review by mimir-d — APPROVED** [review]
thanks for addressing all of the comments; i know this is borderline nitpicky but it's important that we have a good baseline with diags at this point in time

File: Dockerfile
Comment (Goshik92) [reply Dockerfile:12]: Moving `Dockerfile` to a subdirectory will move some files such as `requierements.txt` out of its context. Not sure how to solve this problem.

**Pull request merged**
